### ISSUE TITLE
Tracking Issue for RFC 1826:  Change the default URL of doc.rust-lang.org

### ISSUE THREAD

**Issue by steveklabnik** [rust-lang/rust#44687]
This is a tracking issue for the RFC 1826 (rust-lang/rfcs#1826).

**Steps:**

- [ ] Implement the RFC 
- [ ] Switch doc.rust-lang.org over, which is sort of "stabilization" in this case.

**Unresolved questions:**

No official ones, but there will undoubtedly be tweaks during implementation.

NB: a prototype is at https://rust-docs.herokuapp.com, source here: https://github.com/steveklabnik/rust-docs This can serve as a starting point for discussing said details, but does not literally have to be the basis of implementation.

**Comment by pornel** [comment]
The prototype behaves like `<frameset>` without exposing useful URLs. That's very problematic for me, since I can't copy/share/bookmark documentation pages. Hopefully the final version will have top-level URLs?

**Comment by steveklabnik** [comment]
> I hope that lack of user-visible URLs is just a rough edge of the prototype rather than part of the solution

Yes, absolutely.

**Comment by kzys** [comment]
Seems we already have https://doc.rust-lang.org/1.29.0/, https://doc.rust-lang.org/1.28.0/, ... The only missing part is having a small header on each docs, which can be implemented as a small JavaScript file.

How we can update the JavaScript file? We may be able to generate that from rust-lang/rust's nightly (assuming that the latest nightly includes all stable releases), or have a separate repository.

**Comment by kzys** [comment]
How about something like that?

http://rust-pull-request.s3-website-us-east-1.amazonaws.com/44687/std/

- Version picker is in the page itself, inserted by JavaScript (no iframe)
- The picker is sticky, always shown since Rust is evolving fast still. 
- The version information is hosted as a static JSON file on doc.rust-lang.org (not implemented)

**Comment by steveklabnik** [comment]
Hey @kzys , thanks for poking at this!

What *exactly* does

> Version picker is in the page itself, inserted by JavaScript

mean? *How* is it inserted by JS?

**Comment by kzys** [comment]
I meant, having all versions links on rustdoc's HTML files would not work,
since we want to have links to older docs to newer docs (and vice versa, of
course). We would not want to update all docs on every release.

So, the implementation would need

1. a JSON file that lists all releases (like thanks.rust-lang.org)
2. a small JavaScript to read the JSON file, construct a drop-down.
3. include the JavaScript file from rustdoc's HTML to show the dropdown if
the location is docs.rust-lang.org

Not sure sure where would be the best place to generate 1. 2 and 3 could be
implemented as a part of rustdoc.

**Comment by steveklabnik** [comment]
Right, so the issue here is:

> include the JavaScript file from rustdoc's HTML 

We have already produced the HTML for those old releases, and they can't change. So this approach would work going forward, but wouldn't work for releases previous to that.

**Comment by steveklabnik** [comment]
>  That's not ideal though.

Yeah, we have a hard constraint on not being able to do this.

> "for older releases, please see ..." 

The issue with that is that it doesn't solve one of the primary motivations for doing this: people seeing older versions and not realizing they're out of date.

**Comment by kzys** [comment]
If changing the old docs is not possible, the remaining options are;

- Dynamically replace docs's certain parts via the http server (like https://httpd.apache.org/docs/2.4/mod/mod_substitute.html or http://nginx.org/en/docs/http/ngx_http_sub_module.html), not so sure what we are using though.
- frame/iframe

Would you mind if I ask the reason of not changing the old docs? I agree we shouldn't do that daily/monthly basis, but what if the change is only once?

**Comment by steveklabnik** [comment]
Reproducibility is important, and so releases are immutable.

I’m starting to think the server side idea is the best option...

**Comment by kzys** [comment]
Thanks! That make sense.

Does doc.rust-lang.org use S3 + CloudFront? Not so sure since the old issues sometimes mention "rewriting". On CloudFront, we may be able to use [Lambda@Edge](https://docs.aws.amazon.com/lambda/latest/dg/lambda-edge.html) instead of spinning up a http server.

(Disclaimer - I work for Amazon, but not AWS)

**Comment by steveklabnik** [comment]
>  doc.rust-lang.org use S3 + CloudFront according to the HTTP headers?

Yep! 

> On CloudFront, we may be able to use Lambda@Edge instead of spinning up a http server.

That is an intriguing idea!

**Comment by Mark-Simulacrum** [comment]
FWIW, I think fixing this by including an empty just file from some common location or path that we can update later (e.g. rust-lang.org/doc-1.23.0.js) would be good. The other side of this is I think changing old HTML files uniformly isn't a bad thing: for one thing it's cheaper than any other solution from a monetary perspective and I don't quite see why it would pose too large a problem. We're not changing static.r-l.o artifacts which I think are far more critical to keep constant vs. docs.r-l.o artifacts...

**Comment by steveklabnik** [comment]
> We're not changing static.r-l.o artifacts 

Oh? I thought they were identical. If this is *not* the case, then I feel like there's more wiggle room.

**Comment by Mark-Simulacrum** [comment]
Yes, the docs are served from a different bucket in s3, I'm fairly sure, and if they weren't, we could "easily" start doing so.

**Comment by steveklabnik** [comment]
Okay, then maybe that is the path forward, then.

**Comment by kzys** [comment]
Is it easy to have a file inside the bucket, like
docs.rust-lang.org/common/footer.js ?

If so, we can create the file first, reference from the nigltly docs by
changing rustdoc, and replace all old docs later to include the file.

In addition to that, if we have a specific update script for the bucket, we
probably don't have to change rustdoc.

**Comment by dustinknopoff** [comment]
I'd like to take a stab at this. I'm a little unclear as to where to start from.

**Comment by kzys** [comment]
Thanks @dustinknopoff!

I'm working on an actual prototype based on the mock. You can try that by copy-and-paste the below JavaScript from browser's console.

```
s=document.createElement('script');s.type='text/javascript';s.src='https://blog.8-p.info/en/tmp/2019/rustdoc-version-selector.js';document.body.appendChild(s);
```

The script only supports doc.rust-lang.org's landing page such as https://doc.rust-lang.org/1.31.0/ so far.

My thoughts;
- While reusing the existing version information looks nice, the coupling between this script and the DOM structure is a bit concerning, in terms of maintenance.
- What do we want to have in the combo box? @steveklabnik's initial prototype only has stable, beta, nightly. https://docs.python.org/3/ has 5 versions and I don't know why these five. My prototype shows all releases as is, which may be confusing.

**Comment by QuietMisdreavus** [comment]
For the people subscribed to this issue: @GuillaumeGomez made a crack at it in https://github.com/rust-lang/rust/pull/58603, though it may be different than what was being planned here.

**Comment by GuillaumeGomez** [comment]
Completely forgot about this one actually. :-.

**Comment by jackh726** [comment]
I'm going to go ahead and assign this to @rust-lang/infra, but it might make more sense under T-rustdoc?

**Comment by GuillaumeGomez** [comment]
It's both so it's a good idea to include them as well.
